test/lit/passes/local-cse_all-features.wast

```diff
@@ -0,0 +1,272 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $non-nullable-value (param $ref (ref $A))
+    ;; The value that is repeated is non-nullable, which we must do some fixups
+    ;; for after creating a local of that type.
+    (drop
+      (select (result (ref $A))
+        (local.get $ref)
+        (local.get $ref)
+        (i32.const 1)
+      )
+    )
+    (drop
+      (select (result (ref $A))
+        (local.get $ref)
+        (local.get $ref)
+        (i32.const 1)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $creations
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new_with_rtt $A
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:    (rtt.canon $A)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new_with_rtt $A
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:    (rtt.canon $A)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (array.new_with_rtt $B
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:    (rtt.canon $B)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (array.new_with_rtt $B
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:    (rtt.canon $B)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $creations
+    ;; Allocating GC data has no side effects, but each allocation is unique
+    ;; and so we cannot replace separate allocations with a single one.
+    (drop
+      (struct.new_with_rtt $A
+        (i32.const 1)
+        (rtt.canon $A)
+      )
+    )
+    (drop
+      (struct.new_with_rtt $A
+        (i32.const 1)
+        (rtt.canon $A)
+      )
+    )
+    (drop
+      (array.new_with_rtt $B
+        (i32.const 1)
+        (i32.const 1)
+        (rtt.canon $B)
+      )
+    )
+    (drop
+      (array.new_with_rtt $B
+        (i32.const 1)
+        (i32.const 1)
+        (rtt.canon $B)
+      )
+    )
+  )
+)
+
+(module
+  ;; Real-world testcase from AssemblyScript, containing multiple nested things
+  ;; that can be optimized. The inputs to the add (the xors) are identical, and
+  ;; we can avoid repeating them.
+  ;; CHECK:      (type $i32_i32_=>_i32 (func (param i32 i32) (result i32)))
+
+  ;; CHECK:      (func $div16_internal (param $0 i32) (param $1 i32) (result i32)
+  ;; CHECK-NEXT:  (local $2 i32)
+  ;; CHECK-NEXT:  (i32.add
+  ;; CHECK-NEXT:   (local.tee $2
+  ;; CHECK-NEXT:    (i32.xor
+  ;; CHECK-NEXT:     (i32.shr_s
+  ;; CHECK-NEXT:      (i32.shl
+  ;; CHECK-NEXT:       (local.get $0)
+  ;; CHECK-NEXT:       (i32.const 16)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (i32.const 16)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (i32.shr_s
+  ;; CHECK-NEXT:      (i32.shl
+  ;; CHECK-NEXT:       (local.get $1)
+  ;; CHECK-NEXT:       (i32.const 16)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:      (i32.const 16)
+  ;; CHECK-NEXT:     )
```

Comment:
These two look the same and could have been optimized?

Reply:
Oh sorry, the local numbers are different. But then this test contains only one thing to optimize (as the `CHECK` lines show), while the test description says "containing multiple nested things that can be optimized". Is this what you intended?

Reply:
The inputs to the xors are identical, but that includes their children being identical. That's what it means by "multiple nested things", but maybe that's not very clear... Anyhow, the test checks we only CSE the outer parent, and do not CSE children unnecessarily, and also do not CSE where the locals are different.

Maybe this test is slightly redundant with the above ones, but it's real-world code that is nice to see is optimized 100% right I think, and it existed before in the test suite, this is just a movement.